enhancements/update/standalone-worker-surge.md

```diff
@@ -0,0 +1,295 @@
+to the consumers of that service.
+
+With appropriate cluster management, node draining can be used to ensure that sufficient pods are running to satisfy workload requirements
+at all times - even during updates. "Appropriate cluster management," though, is a multi-faceted challenge involving considerations
+from pod replicas to cluster topology.
+
+### Managing Worker Node Capacity
+
+One aspect of this challenge is ensuring that, while a node is being drained, there is sufficient worker node capacity (CPU/memory/other
+resources) available for new pods to take the place of old pods from the node being drained. Consider the reductive example of 
+a static cluster with a single worker node. If there is an attempt to drain the node in this example, there is no additional worker 
+node capacity available to schedule new pods to replace the pods being drained. This can result in a stalled drain -- one that 
+does not terminate until there is external intervention.
+
+Stalled drains create a frustrating experience for operations teams as they require analysis and intervention. They can also
+make it impossible to predict when an update will complete -- complicating work schedules and communications. There are a number of reasons 
+drains can stall, but simple lack of spare worker node capacity is a common one. The reason is that running an over-provisioned cluster
+(i.e. with more worker nodes than actually required for the steady state workloads) is not always cost-effective. Running
+an extra node 24x7 simply to ensure capacity for the short period while another node is being drained is inefficient.
+
+One cost-effective approach to ensure capacity is called "surging". With a surge strategy, during an update, the platform
+is permitted to bring additional worker nodes online, to accommodate workloads being drained from existing nodes. After an 
+update concludes, the surged nodes are scaled down and the cluster resumes its steady state.
+
+HyperShfit Hosted Control Planes (HCP) already support the surge concept. HyperShift `NodePools`
+expose `maxUnavailable` and `maxSurge` as configurable options during updates: https://hypershift-docs.netlify.app/reference/api/#hypershift.openshift.io/v1beta1.RollingUpdate . 
+Unfortunately, standalone OpenShift, which uses `MachineConfigPools`, does not. To workaround this
+limitation for managed services customers, Service Delivery developed a custom Machine Update Operator (MUO)
+which can surge a standalone cluster during an update (see [reserved capacity feature](https://github.com/openshift/managed-upgrade-operator/blob/a56079fda6ab4088f350b05ed007896a4cabcd97/docs/faq.md)).
+
+### Preventing Other Stalled Drains
+
+As previously mentioned, there are other reasons that drains can stall. For example, `PodDisruptionBudgets` can
+be configured in such a way as to prevent pods from draining even if there is sufficient capacity for them

```

Comment:
When self managed it's fine for the admin to handle this on a case by case basis.  For RH managed openshift it cannot scale, we have to automate this.  And we don't know customer workloads, actually cannot even see them.  Best (only) thing we can do is ensure the upgrade progresses after a reasonable and documented waiting period.

Reply:
I know I'm about to cross into a whole other territory here, but bear with me.

IIUC, the other enhancement currently is talking about separating control plane and worker upgrades in a way such that our end users have greater control over when they are disrupted right? I'm wondering if the "policy" for force upgrading should be a part of that configuration?

Users can't run workloads on the control plane right so I don't think this is a control plane issue.

Once the other EP is merged, how much do we care about workload nodes being upgraded, how much drift can there be without it causing headaches?

Reply:
Customers can set the timeout for forcing drain today at a cluster level.  For drift we care about infra nodes, which managed openshift does deploy.  Those should not fall behind.  Should be a max allowed skew on control plane upgrade.  I'll check the other EP for both of these..

Reply:
I separated the enhancements because I think they have value in isolation, but it is most impactful for SD if we execute on both (i.e. they can dramatically reduce the MUO). I would still suggest separate configuration stanzas:
1. It allows this enhancement config stanza to mirror HCP and MachineDeployment.
2. It is valid and simplifying to separate (a) **when** an update should be performed (https://github.com/openshift/enhancements/pull/1571) and (b) **how** an update should be performed (this enhancement). 
3. No one wants the maintenance schedule enhancement to be even longer :) .

Reply:
> Once the other EP is merged, how much do we care about workload nodes being upgraded.
Much less. I expect most customers will update worker nodes less often. 

> how much drift can there be without it causing headaches?
N-2 today and N-3 tomorrow when we adopt upstream kube's posture.

Reply:
I agree it makes sense to separate the two, but I'm wondering if this becomes much simpler if we assume that the maintenance schedules already exist (ie doing that one first and implementing this after)

> N-2 today and N-3 tomorrow when we adopt upstream kube's posture.

So realistically a maximum drain timeout could be set to a year and not cause an issue, interesting 🤔